File: samples/RefusedBequestExample.java
package com.example.samples;

import java.util.List;
import java.util.ArrayList;

// --- DynamicArray and SimpleStack ---

abstract class DynamicArray {
    protected Object[] elements;
    protected int size;

    public void add(Object element) {
        ensureCapacity();
        elements[size++] = element;
    }

    public Object get(int index) {
        if (index < 0 || index >= size) throw new IndexOutOfBoundsException();
        return elements[index];
    }

    public void set(int index, Object element) {
        if (index < 0 || index >= size) throw new IndexOutOfBoundsException();
        elements[index] = element;
    }

    public void insertAt(int index, Object element) {
        ensureCapacity();
        System.arraycopy(elements, index, elements, index + 1, size - index);
        elements[index] = element;
        size++;
    }

    public Object removeAt(int index) {
        Object old = elements[index];
        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
        size--;
        return old;
    }

    public int size() { return size; }

    public int indexOf(Object o) {
        for (int i = 0; i < size; i++) {
            if (elements[i].equals(o)) return i;
        }
        return -1;
    }

    public boolean contains(Object o) { return indexOf(o) >= 0; }

    public void clear() { size = 0; }

    protected void ensureCapacity() {
        if (elements == null) elements = new Object[10];
        if (size >= elements.length) {
            Object[] bigger = new Object[elements.length * 2];
            System.arraycopy(elements, 0, bigger, 0, size);
            elements = bigger;
        }
    }
}

/** SimpleStack — LIFO wrapper backed by DynamicArray */
public class SimpleStack extends DynamicArray {

    public void push(Object item) {
        add(item);
    }

    public Object pop() {
        if (size == 0) throw new RuntimeException("Stack is empty");
        return removeAt(size - 1);
    }

    public Object peek() {
        if (size == 0) throw new RuntimeException("Stack is empty");
        return elements[size - 1];
    }

    public boolean isEmpty() {
        return size == 0;
    }
}


// --- AbstractReportGenerator and DataExporter ---

abstract class AbstractReportGenerator {
    protected String title;
    protected String author;
    protected List<String> sections;

    public AbstractReportGenerator(String title, String author) {
        this.title = title;
        this.author = author;
        this.sections = new ArrayList<>();
    }

    public abstract String generateHeader();
    public abstract String generateFooter();
    public abstract String formatSection(String content);
    public abstract void addTableOfContents();
    public abstract void addPageNumbers();
    public abstract String renderToHTML();
    public abstract String renderToPDF();

    public void addSection(String section) {
        sections.add(section);
    }

    public String getTitle() { return title; }
    public String getAuthor() { return author; }
}

/** DataExporter — CSV export backed by AbstractReportGenerator */
class DataExporter extends AbstractReportGenerator {
    private String delimiter;

    public DataExporter(String title, String delimiter) {
        super(title, "system");
        this.delimiter = delimiter;
    }

    @Override
    public String generateHeader() { return ""; }

    @Override
    public String generateFooter() { return ""; }

    @Override
    public String formatSection(String content) { return content; }

    @Override
    public void addTableOfContents() { }

    @Override
    public void addPageNumbers() { }

    @Override
    public String renderToHTML() { return ""; }

    @Override
    public String renderToPDF() { return ""; }

    public String exportToCSV(List<String[]> data) {
        StringBuilder sb = new StringBuilder();
        for (String[] row : data) {
            sb.append(String.join(delimiter, row)).append("\n");
        }
        return sb.toString();
    }

    public void exportToFile(List<String[]> data, String filePath) {
        String csv = exportToCSV(data);
        // Write to file...
    }
}

File: samples js/RefusedBequestExample.js
const fs = require("fs");

// --- DynamicArray and SimpleStack ---

class DynamicArray {
    constructor() {
        this.elements = [];
    }

    add(element) {
        this.elements.push(element);
    }

    get(index) {
        if (index < 0 || index >= this.elements.length) throw new RangeError("Index out of bounds");
        return this.elements[index];
    }

    set(index, element) {
        if (index < 0 || index >= this.elements.length) throw new RangeError("Index out of bounds");
        this.elements[index] = element;
    }

    insertAt(index, element) {
        this.elements.splice(index, 0, element);
    }

    removeAt(index) {
        return this.elements.splice(index, 1)[0];
    }

    size() {
        return this.elements.length;
    }

    indexOf(o) {
        return this.elements.indexOf(o);
    }

    contains(o) {
        return this.indexOf(o) >= 0;
    }

    clear() {
        this.elements.length = 0;
    }
}

/** SimpleStack — LIFO wrapper backed by DynamicArray */
class SimpleStack extends DynamicArray {
    push(item) {
        this.add(item);
    }

    pop() {
        if (this.elements.length === 0) throw new Error("Stack is empty");
        return this.removeAt(this.elements.length - 1);
    }

    peek() {
        if (this.elements.length === 0) throw new Error("Stack is empty");
        return this.elements[this.elements.length - 1];
    }

    isEmpty() {
        return this.elements.length === 0;
    }
}


// --- AbstractReportGenerator and DataExporter ---

class AbstractReportGenerator {
    constructor(title, author) {
        this.title = title;
        this.author = author;
        this.sections = [];
    }

    generateHeader() { throw new Error("Not implemented"); }
    generateFooter() { throw new Error("Not implemented"); }
    formatSection(content) { throw new Error("Not implemented"); }
    addTableOfContents() { throw new Error("Not implemented"); }
    addPageNumbers() { throw new Error("Not implemented"); }
    renderToHTML() { throw new Error("Not implemented"); }
    renderToPDF() { throw new Error("Not implemented"); }

    addSection(section) {
        this.sections.push(section);
    }
}

/** DataExporter — CSV export backed by AbstractReportGenerator */
class DataExporter extends AbstractReportGenerator {
    constructor(title, delimiter) {
        super(title, "system");
        this.delimiter = delimiter;
    }

    generateHeader() { return ""; }
    generateFooter() { return ""; }
    formatSection(content) { return content; }
    addTableOfContents() { /* no-op */ }
    addPageNumbers() { /* no-op */ }
    renderToHTML() { return ""; }
    renderToPDF() { return ""; }

    exportToCSV(data) {
        return data.map(row => row.join(this.delimiter)).join("\n");
    }

    exportToFile(data, filePath) {
        const csv = this.exportToCSV(data);
        fs.writeFileSync(filePath, csv);
    }
}

module.exports = { DynamicArray, SimpleStack, AbstractReportGenerator, DataExporter };
